// @wyre-technology/node-rocketcyber/dist/index.js
'use strict';

// src/config.ts
var REGION_URLS = {
  us: "https://api-us.rocketcyber.com/v3"
};
var DEFAULT_RATE_LIMIT_CONFIG = {
  enabled: true,
  maxRequests: 60,
  windowMs: 6e4,
  throttleThreshold: 0.8,
  retryAfterMs: 5e3,
  maxRetries: 3
};
function resolveConfig(config) {
  if (!config.apiKey) {
    throw new Error("apiKey is required");
  }
  let baseUrl;
  if (config.baseUrl) {
    baseUrl = config.baseUrl.replace(/\/$/, "");
  } else {
    const region = config.region ?? "us";
    const regionUrl = REGION_URLS[region];
    if (!regionUrl) {
      throw new Error(`Invalid region: ${region}. Valid regions are: us`);
    }
    baseUrl = regionUrl;
  }
  return {
    apiKey: config.apiKey,
    baseUrl,
    rateLimit: {
      ...DEFAULT_RATE_LIMIT_CONFIG,
      ...config.rateLimit
    }
  };
}

// src/auth.ts
var AuthManager = class {
  config;
  constructor(config) {
    this.config = config;
  }
  getAuthHeaders() {
    return {
      "Authorization": `Bearer ${this.config.apiKey}`
    };
  }
};

// src/errors.ts
var RocketCyberError = class _RocketCyberError extends Error {
  statusCode;
  response;
  constructor(message, statusCode = 0, response) {
    super(message);
    this.name = "RocketCyberError";
    this.statusCode = statusCode;
    this.response = response;
    Object.setPrototypeOf(this, _RocketCyberError.prototype);
  }
};
var RocketCyberAuthenticationError = class _RocketCyberAuthenticationError extends RocketCyberError {
  constructor(message, statusCode = 401, response) {
    super(message, statusCode, response);
    this.name = "RocketCyberAuthenticationError";
    Object.setPrototypeOf(this, _RocketCyberAuthenticationError.prototype);
  }
};
var RocketCyberForbiddenError = class _RocketCyberForbiddenError extends RocketCyberError {
  constructor(message, response) {
    super(message, 403, response);
    this.name = "RocketCyberForbiddenError";
    Object.setPrototypeOf(this, _RocketCyberForbiddenError.prototype);
  }
};
var RocketCyberNotFoundError = class _RocketCyberNotFoundError extends RocketCyberError {
  constructor(message, response) {
    super(message, 404, response);
    this.name = "RocketCyberNotFoundError";
    Object.setPrototypeOf(this, _RocketCyberNotFoundError.prototype);
  }
};
var RocketCyberRateLimitError = class _RocketCyberRateLimitError extends RocketCyberError {
  retryAfter;
  constructor(message, retryAfter = 5e3, response) {
    super(message, 429, response);
    this.name = "RocketCyberRateLimitError";
    this.retryAfter = retryAfter;
    Object.setPrototypeOf(this, _RocketCyberRateLimitError.prototype);
  }
};
var RocketCyberServerError = class _RocketCyberServerError extends RocketCyberError {
  constructor(message, statusCode = 500, response) {
    super(message, statusCode, response);
    this.name = "RocketCyberServerError";
    Object.setPrototypeOf(this, _RocketCyberServerError.prototype);
  }
};

// src/http.ts
var HttpClient = class {
  config;
  authManager;
  rateLimiter;
  constructor(config, authManager, rateLimiter) {
    this.config = config;
    this.authManager = authManager;
    this.rateLimiter = rateLimiter;
  }
  async request(path, options = {}) {
    const { method = "GET", body, params } = options;
    let url = `${this.config.baseUrl}${path}`;
    if (params) {
      const searchParams = new URLSearchParams();
      for (const [key, value] of Object.entries(params)) {
        if (value !== void 0) {
          searchParams.append(key, String(value));
        }
      }
      const queryString = searchParams.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }
    return this.executeRequest(url, method, body);
  }
  async requestUrl(url) {
    return this.executeRequest(url, "GET", void 0);
  }
  async executeRequest(url, method, body, retryCount = 0) {
    await this.rateLimiter.waitForSlot();
    const headers = {
      "Content-Type": "application/json",
      ...this.authManager.getAuthHeaders()
    };
    this.rateLimiter.recordRequest();
    const response = await fetch(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : void 0
    });
    return this.handleResponse(response, url, method, body, retryCount);
  }
  async handleResponse(response, url, method, body, retryCount) {
    if (response.ok) {
      if (response.status === 204) {
        return {};
      }
      const contentType = response.headers.get("content-type");
      if (contentType?.includes("application/json")) {
        return response.json();
      }
      return {};
    }
    let responseBody;
    try {
      responseBody = await response.json();
    } catch {
      responseBody = await response.text();
    }
    switch (response.status) {
      case 400:
        throw new RocketCyberError("Bad request", 400, responseBody);
      case 401:
        throw new RocketCyberAuthenticationError(
          "Authentication failed - invalid API key",
          401,
          responseBody
        );
      case 403:
        throw new RocketCyberForbiddenError("Access forbidden - insufficient permissions", responseBody);
      case 404:
        throw new RocketCyberNotFoundError("Resource not found", responseBody);
      case 429:
        if (this.rateLimiter.shouldRetry(retryCount)) {
          const retryAfterHeader = response.headers.get("Retry-After");
          const delay = this.rateLimiter.parseRetryAfter(retryAfterHeader);
          await this.sleep(delay);
          return this.executeRequest(url, method, body, retryCount + 1);
        }
        throw new RocketCyberRateLimitError(
          "Rate limit exceeded and max retries reached",
          this.config.rateLimit.retryAfterMs,
          responseBody
        );
      default:
        if (response.status >= 500) {
          if (retryCount === 0) {
            await this.sleep(1e3);
            return this.executeRequest(url, method, body, 1);
          }
          throw new RocketCyberServerError(
            `Server error: ${response.status} ${response.statusText}`,
            response.status,
            responseBody
          );
        }
        throw new RocketCyberError(
          `Request failed: ${response.status} ${response.statusText}`,
          response.status,
          responseBody
        );
    }
  }
  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
};

// src/rate-limiter.ts
var RateLimiter = class {
  config;
  requestTimestamps = [];
  constructor(config) {
    this.config = config;
  }
  /**
   * Wait until it's safe to make a request
   */
  async waitForSlot() {
    if (!this.config.enabled) {
      return;
    }
    this.pruneOldTimestamps();
    const currentRate = this.requestTimestamps.length / this.config.maxRequests;
    if (currentRate >= this.config.throttleThreshold) {
      const delayMs = Math.min(
        1e3 * (currentRate - this.config.throttleThreshold + 0.1) * 10,
        5e3
      );
      await this.sleep(delayMs);
    }
    if (this.requestTimestamps.length >= this.config.maxRequests) {
      const oldestTimestamp = this.requestTimestamps[0];
      if (oldestTimestamp !== void 0) {
        const waitUntil = oldestTimestamp + this.config.windowMs;
        const waitTime = waitUntil - Date.now();
        if (waitTime > 0) {
          await this.sleep(waitTime);
        }
      }
    }
  }
  /**
   * Record that a request was made
   */
  recordRequest() {
    if (!this.config.enabled) {
      return;
    }
    this.requestTimestamps.push(Date.now());
  }
  /**
   * Get the current request rate as a fraction of the limit
   */
  getCurrentRate() {
    this.pruneOldTimestamps();
    return this.requestTimestamps.length / this.config.maxRequests;
  }
  /**
   * Get the number of requests remaining in the current window
   */
  getRemainingRequests() {
    this.pruneOldTimestamps();
    return Math.max(0, this.config.maxRequests - this.requestTimestamps.length);
  }
  /**
   * Calculate retry delay with exponential backoff
   */
  calculateRetryDelay(retryCount) {
    return Math.min(
      this.config.retryAfterMs * Math.pow(2, retryCount),
      3e4
      // Max 30 seconds
    );
  }
  /**
   * Check if we should retry after a rate limit error
   */
  shouldRetry(retryCount) {
    return retryCount < this.config.maxRetries;
  }
  /**
   * Parse Retry-After header value
   * Returns delay in milliseconds
   */
  parseRetryAfter(retryAfterHeader) {
    if (!retryAfterHeader) {
      return this.config.retryAfterMs;
    }
    const seconds = parseInt(retryAfterHeader, 10);
    if (!isNaN(seconds)) {
      return seconds * 1e3;
    }
    try {
      const date = new Date(retryAfterHeader);
      const delay = date.getTime() - Date.now();
      if (delay > 0) {
        return delay;
      }
    } catch {
    }
    return this.config.retryAfterMs;
  }
  /**
   * Remove timestamps older than the window
   */
  pruneOldTimestamps() {
    const cutoff = Date.now() - this.config.windowMs;
    this.requestTimestamps = this.requestTimestamps.filter((ts) => ts > cutoff);
  }
  /**
   * Sleep for a given duration
   */
  sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
};

// src/resources/account.ts
var AccountResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async get(params) {
    return this.http.request("/account", {
      params
    });
  }
};

// src/resources/agents.ts
var AgentsResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async list(params) {
    return this.http.request("/agents", {
      params
    });
  }
  async listAll(params) {
    const allItems = [];
    let page = 1;
    while (true) {
      const response = await this.list({ ...params, page });
      allItems.push(...response.data);
      if (response.currentPage >= response.totalPages) break;
      page++;
    }
    return allItems;
  }
};

// src/resources/incidents.ts
var IncidentsResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async list(params) {
    return this.http.request("/incidents", {
      params
    });
  }
  async listAll(params) {
    const allItems = [];
    let page = 1;
    while (true) {
      const response = await this.list({ ...params, page });
      allItems.push(...response.data);
      if (response.currentPage >= response.totalPages) break;
      page++;
    }
    return allItems;
  }
};

// src/resources/events.ts
var EventsResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async list(params) {
    return this.http.request("/events", {
      params
    });
  }
  async listAll(params) {
    const allItems = [];
    let page = 1;
    while (true) {
      const response = await this.list({ ...params, page });
      allItems.push(...response.data);
      if (response.currentPage >= response.totalPages) break;
      page++;
    }
    return allItems;
  }
  async getSummary(params) {
    return this.http.request("/events/summary", {
      params
    });
  }
};

// src/resources/firewalls.ts
var FirewallsResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async list(params) {
    return this.http.request("/firewalls", {
      params
    });
  }
  async listAll(params) {
    const allItems = [];
    let page = 1;
    while (true) {
      const response = await this.list({ ...params, page });
      allItems.push(...response.data);
      if (response.currentPage >= response.totalPages) break;
      page++;
    }
    return allItems;
  }
};

// src/resources/apps.ts
var AppsResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async list(params) {
    return this.http.request("/apps", {
      params
    });
  }
  async listAll(params) {
    const allItems = [];
    let page = 1;
    while (true) {
      const response = await this.list({ ...params, page });
      allItems.push(...response.data);
      if (response.currentPage >= response.totalPages) break;
      page++;
    }
    return allItems;
  }
};

// src/resources/defender.ts
var DefenderResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async get(params) {
    return this.http.request("/defender", {
      params
    });
  }
};

// src/resources/office.ts
var OfficeResource = class {
  http;
  constructor(http) {
    this.http = http;
  }
  async get(params) {
    return this.http.request("/office", {
      params
    });
  }
};

// src/client.ts
var RocketCyberClient = class {
  config;
  authManager;
  rateLimiter;
  httpClient;
  account;
  agents;
  incidents;
  events;
  firewalls;
  apps;
  defender;
  office;
  constructor(config) {
    this.config = resolveConfig(config);
    this.authManager = new AuthManager(this.config);
    this.rateLimiter = new RateLimiter(this.config.rateLimit);
    this.httpClient = new HttpClient(this.config, this.authManager, this.rateLimiter);
    this.account = new AccountResource(this.httpClient);
    this.agents = new AgentsResource(this.httpClient);
    this.incidents = new IncidentsResource(this.httpClient);
    this.events = new EventsResource(this.httpClient);
    this.firewalls = new FirewallsResource(this.httpClient);
    this.apps = new AppsResource(this.httpClient);
    this.defender = new DefenderResource(this.httpClient);
    this.office = new OfficeResource(this.httpClient);
  }
  getConfig() {
    return this.config;
  }
  getRateLimitStatus() {
    return {
      remaining: this.rateLimiter.getRemainingRequests(),
      rate: this.rateLimiter.getCurrentRate()
    };
  }
};

exports.AccountResource = AccountResource;
exports.AgentsResource = AgentsResource;
exports.AppsResource = AppsResource;
exports.DEFAULT_RATE_LIMIT_CONFIG = DEFAULT_RATE_LIMIT_CONFIG;
exports.DefenderResource = DefenderResource;
exports.EventsResource = EventsResource;
exports.FirewallsResource = FirewallsResource;
exports.IncidentsResource = IncidentsResource;
exports.OfficeResource = OfficeResource;
exports.REGION_URLS = REGION_URLS;
exports.RocketCyberAuthenticationError = RocketCyberAuthenticationError;
exports.RocketCyberClient = RocketCyberClient;
exports.RocketCyberError = RocketCyberError;
exports.RocketCyberForbiddenError = RocketCyberForbiddenError;
exports.RocketCyberNotFoundError = RocketCyberNotFoundError;
exports.RocketCyberRateLimitError = RocketCyberRateLimitError;
exports.RocketCyberServerError = RocketCyberServerError;
exports.resolveConfig = resolveConfig;
